### PR TITLE
Feature/quorum handler

### PR DESCRIPTION
**Pull request by DenisRybas**
Added quorum handler and tests for it for read and write transaction

### CODE REVIEW

File: indy-besu/vdr/Cargo.toml [hunk 1]
```diff
@@ -21,6 +21,9 @@ basic_signer = ["secp256k1"]
 migration = []
 
 [dependencies]
+tokio = "1.34.0"
+futures-core = "0.3"
+futures-util = "0.3"
```
Comment (Artemkaaas) [inline indy-besu/vdr/Cargo.toml]: Are these dependencies (`futures-core` and `futures-util` ) really need to be listed here?

File: indy-besu/vdr/Cargo.toml [hunk 2]
```diff
@@ -30,16 +33,24 @@ ethabi = "18.0.0"
 hex = "0.4.3"
 futures = "0.3.28"
 once_cell = "1.18.0"
-secp256k1 = { version = "0.28.0", optional = true, features = ["recovery", "rand"] }
+secp256k1 = { version = "0.28.0", optional = true, features = [
```
Comment (Artemkaaas) [inline indy-besu/vdr/Cargo.toml]: such formatting looks strange

File: indy-besu/vdr/src/client/client.rs [hunk 3]
```diff
@@ -30,27 +33,40 @@ impl LedgerClient {
     ///  client to use for building and sending transactions
     pub fn new(
         chain_id: u64,
-        node_address: &str,
+        mut node_addresses: Vec<&str>,
```
Comment (Artemkaaas) [inline indy-besu/vdr/src/client/client.rs]: Do not like such API at all 

```
mut node_addresses
let rpc_node_address = node_addresses.remove(0);
```

Comment (Artemkaaas) [inline indy-besu/vdr/src/client/client.rs]: Should we better pass `quorum_config` to the client creation?

File: indy-besu/vdr/src/client/client.rs [hunk 4]
```diff
@@ -30,27 +33,40 @@ impl LedgerClient {
...
         contract_configs: &[ContractConfig],
     ) -> VdrResult<LedgerClient> {
+        let rpc_node_address = node_addresses.remove(0);
```
Comment (Artemkaaas) [inline indy-besu/vdr/src/client/client.rs]: If we still need to pass RPC node it will be better to have such API:
```
    pub fn new(
        chain_id: u64,
        rpc_node: &str,
        contracts: &[ContractConfig],
        quorum_config: Option<&QuorumConfig>,
    ) -> VdrResult<LedgerClient> {
```
where QuorumConfig
```
pub struct QuorumConfig {
    nodes: Vec<&str>,
    request_retries: u8,
    request_timeout: u8,
}
```
and ledger client will store quorum handle if config passed and valid 
request quorum if ihandler was initialized

File: indy-besu/vdr/src/client/client.rs [hunk 5]
```diff
@@ -30,27 +33,40 @@ impl LedgerClient {
...
+                let client: Box<dyn Client> = Box::new(Web3Client::new(node_address)?);
+                Ok(Arc::new(client))
+            })
+            .collect::<Result<Vec<_>, VdrError>>()
+            .map_err(|_| VdrError::ClientNodeUnreachable)?;
+
+        let client = Box::new(Web3Client::new(rpc_node_address)?);
         let contracts = Self::init_contracts(&client, contract_configs)?;
 
+        clients.insert(0, Arc::new(client));
```
Comment (Artemkaaas) [inline indy-besu/vdr/src/client/client.rs]: I think RPC node should not participate in quorum check

File: indy-besu/vdr/src/client/client.rs [hunk 6]
```diff
@@ -30,27 +33,40 @@ impl LedgerClient {
...
+
         trace!(
             "Started creating new LedgerClient. Chain id: {}, node address: {}",
             chain_id,
-            node_address
+            rpc_node_address
         );
 
-        let client = Web3Client::new(node_address)?;
+        let mut clients = node_addresses
```
Comment (Artemkaaas) [inline indy-besu/vdr/src/client/client.rs]: These clients should be stored in Quorum Handler

File: indy-besu/vdr/src/client/client.rs [hunk 7]
```diff
@@ -61,7 +77,7 @@ impl LedgerClient {
     /// # Returns
     ///  ping status
     pub async fn ping(&self) -> VdrResult<PingStatus> {
-        self.client.ping().await
+        self.clients[0].ping().await
```
Comment (Artemkaaas) [inline indy-besu/vdr/src/client/client.rs]: rpc node should be used here

File: indy-besu/vdr/src/client/client.rs [hunk 8]
```diff
@@ -73,10 +89,32 @@ impl LedgerClient {
...
+        transaction: &Transaction,
+        quorum_config: &QuorumConfig,
+    ) -> VdrResult<Vec<u8>> {
         match transaction.type_ {
-            TransactionType::Read => self.client.call_transaction(transaction).await,
-            TransactionType::Write => self.client.submit_transaction(transaction).await,
+            TransactionType::Read => {
+                let call_result = self.clients[0].call_transaction(transaction).await?;
+
+                read_quorum::quorum_check(&self.clients, transaction, &call_result, quorum_config)
```
Comment (Artemkaaas) [inline indy-besu/vdr/src/client/client.rs]:
```
let result = match transaction.type_ {
            TransactionType::Read => self.client.call_transaction(transaction).await,
            TransactionType::Write => self.client.submit_transaction(transaction).await,
}
if let Some(qurum_handler) = self.qurum_handler {
   qurum_handler.check(transaction, result)
}
```

File: indy-besu/vdr/src/client/implementation/web3/client.rs [hunk 9]
```diff
@@ -54,7 +55,7 @@ impl Web3Client {
     }
 }
 
-#[async_trait(?Send)]
+#[async_trait::async_trait]
```
Comment (Artemkaaas) [inline indy-besu/vdr/src/client/implementation/web3/client.rs]: Why did you change it? 
`#[async_trait(?Send)]` required for wasm compilation.

File: indy-besu/vdr/src/client/mod.rs [hunk 10]
```diff
@@ -9,11 +10,16 @@ use crate::{
...
+use web3::types::{Transaction as Web3Transaction, H256};
 
-#[async_trait(?Send)]
-pub trait Client {
+#[cfg(test)]
+use mockall::automock;
+
+#[cfg_attr(test, automock)]
+#[async_trait::async_trait]
+pub trait Client: Send + Sync {
```
Comment (Artemkaaas) [inline indy-besu/vdr/src/client/mod.rs]: return `#[async_trait(?Send)]` or put udner wasm feature

File: indy-besu/vdr/src/client/client.rs (hunk 3 above)
Comment (Artemkaaas) [reply indy-besu/vdr/src/client/client.rs]: Update function description

File: indy-besu/vdr/src/client/quorum.rs [hunk 11]
```diff
@@ -0,0 +1,569 @@
...
+                break;
+            } else {
+                trace!(
+                    "eth_getTransaction not succeed for transaction_hash: {:?}. retry",
+                    transaction_hash
+                );
+            }
+        }
+
+        drop(sender);
```
Comment (Artemkaaas) [inline indy-besu/vdr/src/client/quorum.rs]: do we really need to call it explicitly?

File: indy-besu/vdr/src/client/quorum.rs [hunk 12]
```diff
@@ -0,0 +1,569 @@
...
+        );
+
+        for _ in 1..=max_retries_num {
+            if let Ok(Some(transaction)) = client.get_transaction(transaction_hash).await {
+                if let Err(_) = sender.send(transaction).await {
+                    trace!("Receiver is closed for sender: {:?}", sender);
+                }
+                break;
+            } else {
+                trace!(
```
Comment (Artemkaaas) [inline indy-besu/vdr/src/client/quorum.rs]: Should we wait for some time before retry?

File: indy-besu/vdr/src/client/quorum.rs [hunk 13]
```diff
@@ -0,0 +1,569 @@
...
+        } else {
+            trace!("Quorum failed for transaction: {:?}", transaction);
+
+            Err(VdrError::QuorumNotReached(format!(
+                "Quorum not reached for transaction: {:?}",
+                transaction,
+            )))
+        }
+    }
+}
```
Comment (Artemkaaas) [inline indy-besu/vdr/src/client/quorum.rs:208]: Their logic for write and read quorum is very similar. Can we somehow reuse it?

**Review by Artemkaaas — CHANGES_REQUESTED** [review]

File: indy-besu/vdr/Cargo.toml (hunk 1 above)
Comment (DenisRybas) [reply indy-besu/vdr/Cargo.toml]: No. Removed

File: indy-besu/vdr/Cargo.toml (hunk 2 above)
Comment (DenisRybas) [reply indy-besu/vdr/Cargo.toml]: Reformatted

File: indy-besu/vdr/src/client/client.rs (hunk 3 above)
Comment (DenisRybas) [reply indy-besu/vdr/src/client/client.rs]: Changed API

Comment (DenisRybas) [reply indy-besu/vdr/src/client/client.rs]: Updated

File: indy-besu/vdr/src/client/client.rs (hunk 4 above)
Comment (DenisRybas) [reply indy-besu/vdr/src/client/client.rs]: Changed API

File: indy-besu/vdr/src/client/client.rs (hunk 5 above)
Comment (DenisRybas) [reply indy-besu/vdr/src/client/client.rs]: Sure. Removed

File: indy-besu/vdr/src/client/client.rs (hunk 6 above)
Comment (DenisRybas) [reply indy-besu/vdr/src/client/client.rs]: Moved

File: indy-besu/vdr/src/client/client.rs (hunk 7 above)
Comment (DenisRybas) [reply indy-besu/vdr/src/client/client.rs]: Fixed

File: indy-besu/vdr/src/client/quorum.rs (hunk 11 above)
Comment (DenisRybas) [reply indy-besu/vdr/src/client/quorum.rs]: In this case no. Removed

File: indy-besu/vdr/src/client/quorum.rs (hunk 12 above)
Comment (DenisRybas) [reply indy-besu/vdr/src/client/quorum.rs]: Added param for retry interval

File: indy-besu/vdr/src/client/quorum.rs (hunk 13 above)
Comment (DenisRybas) [reply indy-besu/vdr/src/client/quorum.rs:208]: Yes. Updated

**Comment by Artemkaaas** [comment]
@DenisRybas seems like you replied to comments but not pushed the changes

**Review by Artemkaaas — APPROVED** [review]
